### DISCUSSION TITLE
Converting .cr2 / .xmp images

### DISCUSSION THREAD

**Discussion by bmoffat-bp** [ImageMagick/ImageMagick#2552] [Help] [answered]
I am very new to the world of image conversion, and have learned just enough to be able to support the one automated conversion process we use.  Now I am faced with the possibility of converting a client's .cr2 raw images, but have learned that when .cr2 images are adjusted, those adjustments are stored in a .xmp sidecar file.
Does ImageMagick support ..cr2/.xmp conversions (specifically to .jpg)?
Running with the software I have installed to trim white space or scale the image results in a message of no EncodeDelegate for the cr2 image format.

Any guidance / advice would be appreciated.  Again, I apologize for being a noob.

**Comment by fmw42** [comment]
Post an example CR2 image. Also what is your ImageMagick version and platform/OS. Please always provide that information when asking questions, since syntax may vary. ImageMagick can read CR2 images, but I am not sure what information needs to be extracted from the xmp file.

**Reply by bmoffat-bp** [reply]
The version I have of IM on a Windows10 VM:
Version: ImageMagick 7.0.10-23 Q16 x64 2020-07-04 http://www.imagemagick.org
Copyright: Copyright (C) 1999-2018 ImageMagick Studio LLC
License: http://www.imagemagick.org/script/license.php
Visual C++: 192628806
Features: Cipher DPC HDRI OpenCL
Delegates (built-in): bzlib cairo flif freetype gslib heic jng jp2 jpeg lcms lqr lzma openexr pangocairo png ps raw rsvg tiff webp xml zlib

The sample I used came from the internet:  https://filesamples.com/formats/cr2  It is a 64K image so not going to post directly here.

Our original conversion steps were:
1. Trim white space
2. Resize image
3. Convert to .jpg
But steps 1 and 2 complain about no EncodeDelegate, but the process did produce a full (pixel-sized) .jpg, so I run the Convert to .jpg first and it appears to do the right thing - at least in converting the image as it exists in the .cr2 file (without any mods specified in the .xmp sidecar).  The image I downloaded from the net did not come with an .xmp sidecar.

Please see the reply from snibgo, as I think that points to the issue with this client (who does use Adobe to edit .cr2 files, producing .xmp sidecars).

**Comment by snibgo** [comment]
IM can read CR2 files, and write the image to whatever format you want, such as JPG.

IM can't write CR2 files. I'm not aware of any software that can write CR2.

The .xmp file has come from a a raw-image editor such as darktable. It is a text file. If you look at it, it should be obvious what software wrote it.

In general, an xmp file is specific to the software that wrote it. IM doesn't know what to do with the metadata. You should use the same software that wrote the file.

**Reply by bmoffat-bp** [reply]
Thank you both for the quick responses!!  This looks to be the guidance I needed. I will need to work with the client to attempt to automate the conversion of their catalog of .cr2+.xmp images to something we can work with that preserves image integrity.  Thank you!

**Comment by snibgo** [comment, marked as answer]
> ...results in a message of no EncodeDelegate for the cr2 image format.

Please show your exact commands. That message should occur only if you are trying to _write_ (ie encode) a CR2 file, not when you are simply reading (ie decode) CR2.

**Reply by bmoffat-bp** [reply]
In the case of the Trim white space step, the call is:
magick convert -quiet "$($Image.FullName)[0]" -bordercolor white -border 0 -fuzz 1% -trim +repage $Image.FullName

$Image is the path to the file being converted - in this case the .cr2 file.  My observation of behavior is that my process expects the above call to magick to modify-in-place (rewrite the .cr2 file with white space removed), but I gather from earlier responses that IM doesn't write .cr2 format, and this is what generates the No EncodeDelegate message.  This explains why moving the "convert to jpg" step first allows the "trim white space" and "resize image" steps to output .jpg.

**Reply by fmw42** [reply]
If you are using IM 7, the use magick, not magick convert and not convert.  What is the file format for your output file $Image.FullName. If .cr2, then change it to directly save as JPG

**Comment by snibgo** [comment]
You can read the CR2, trim it, resize it etc and write to a JPG in a single magick command. As you now realise, you can't write to a CR2.

**Reply by bmoffat-bp** [reply]
What? Really?  Any chance you could provide the command line to do that?  I know I should go figure it out myself but I have several other plates in the air right now, and I would very much appreciate your guidance to do it right the first time.  Thanks!

**Reply by snibgo** [reply]
For example:
```
magick convert -quiet "$($Image.FullName)[0]" -bordercolor white -border 0 -fuzz 1% -trim +repage -resize 123x456 out.jpg
```
But I have no idea why you are adding a border of zero pixels.

**Reply by bmoffat-bp** [reply]
My only guess is that at some point we played with putting a white border around images to separate them when placed adjacent to each other, and no longer need that due to UI layout taking care of it, and just set it to 0 in case someone changed their minds again?

Thank you for the command line above!!
